src/collective/collectionfilter/query.py

```diff
@@ -7,7 +7,37 @@
 from logging import getLogger
 from zope.component import getUtility
 
+
 logger = getLogger('collective.collectionfilter')
+MULTISPACE = u'\u3000'
+BAD_CHARS = ('?', '-', '+', '*', MULTISPACE)
+
+
+def quote_unsafe_chars(s):
+    # We need to quote parentheses when searching text indices
+    if '(' in s:
+        s = s.replace('(', '"("')
+    if ')' in s:
+        s = s.replace(')', '")"')
+    if MULTISPACE in s:
+        s = s.replace(MULTISPACE, ' ')
+    return s
+
+
+def quote_keywords(term):
+    # The terms and, or and not must be wrapped in quotes to avoid
+    # being parsed as logical query atoms.
+    if term.lower() in ('and', 'or', 'not'):
+        term = '"%s"' % term
+    return term
+
+
+def sanitise_search_query(query):
+    for char in BAD_CHARS:
+        query = query.replace(char, u" ")
+    clean_query = [quote_keywords(token) for token in query.split()]
+    clean_query = quote_unsafe_chars(clean_query)
+    return u" ".join(clean_query)
```

Comment:
There is similar code here https://github.com/plone/Products.CMFPlone/blob/master/Products/CMFPlone/browser/search.py#L23 ... maybe you can reuse it in some way?